### DISCUSSION TITLE
D400 performance

### DISCUSSION THREAD

**Discussion by phoenixflare** [thirtythreeforty/neolink#172] [Q&A]
So after mucking around with this on ubuntu for some time, I finally got a new D400 camera to work out of the box. 

The issue seems to be that the frame rate seems to be around 1 fps every 1 to 5 seconds while it updates about 30 seconds late. So if I was to wave my hand in front of the camera, I'd see my hand blur on the screen about 30 seconds after I waved my hand.

Also I doubt it but there's no way to get rid of the reolink watermark right?

**Comment by phoenixflare** [comment]
Was doing some more testing and it seems like making this into a service increased the performance. To about a 10 second delay with 2fps.

**Comment by QuantumEntangledAndy** [comment]
Not sure whats up with your fps other than maybe bandwidth or computer resources but I believe that the watermark can be disabled in the settings with the official app in the same place you change the other OSD settings

**Reply by phoenixflare** [reply]
Don't have a NVR. Literally got a few of these for testing to see if I can get them to work strictly as a RTSP camera.

The fps could be just the motioneye problem so I guess I'll spin up a shinobi instance to test this and maybe a blueiris after.

I couldn't get the stream to work on VLC either so not sure what is going on there.

**Reply by QuantumEntangledAndy** [reply]
Not sure what the NVR has to do with my comment but ok... For testing I usually recommend ffplay. With VLC you need to either increase the UDP buffer or force it to use TCP.

Here's the OSD settings I was talking about:

<img width="675" alt="Screenshot 2021-06-23 at 12 18 56" src="https://user-images.githubusercontent.com/13386481/123040011-3c8dd400-d41d-11eb-9884-b7b56210b536.png">

**Reply by QuantumEntangledAndy** [reply]
Neolink will let you run it as an RTSP camera by proxying its custom format into an RTSP stream. If you are extra cautious you may want to block the cameras external access to the internet on your router firewall, they have a tendency to phone home and the custom format data stream they use internally is not particularly secure.

**Reply by phoenixflare** [reply]
I thought I needed a nvr for that. Didn't realize the reolink client was all I needed. Disabling everything seemed to have fixed the framerate issue and delay. Well at least it's down to 8 second delay and 30fps.

**Reply by QuantumEntangledAndy** [reply]
When you use the offical client do you also see a delay? With mine I get about 2-3s delay

**Reply by phoenixflare** [reply]
Roughly a 0.5 second delay.
![image](https://user-images.githubusercontent.com/8162563/123044487-94d9bd00-d3ae-11eb-8787-9707e61f187e.png)

**Reply by phoenixflare** [reply]
Ok last update I guess. So apparently there is about a 5 seconds delay when testing using ffplay due to it caching the data so... ya

Properly set up the camera on motioneye with the same video resolution and the delay dropped down to 0.5 seconds. Which is about what the reolink client gets without the neolink in between. Guess the reencoding done by motioneye was killing both the framerate and quality if I tried to use a lower resolution.

For the hell of it I decided to test if this would work with more cpu/ram since I had this running in a virtualbox VM. 
Results: 0.6 seconds delay with 1 core and 1gb of ram and the delay goes up to 1.7 seconds if I used 1 core and 0.5gb of ram. Adding more core and/or ram didn't change the performance at all.

**Reply by QuantumEntangledAndy** [reply]
Interesting  to know about the buffer, is that the UDP buffer by any chance? If you not already maybe you could add `-rtsp_transport tcp` to see.

I wouldn't expect cpus to be doing much for a single camera. We only really empoy multi-threading per camera so one cpu should be fine for one camera. Similarly with RAM neolink just needs enough for the buffer and after that extra ram is mostly unimportant.

**Reply by phoenixflare** [reply]
The command I am using for ffplay is
`ffplay.exe -rtsp_transport tcp rtsp://192.168.1.10:8554/d400`

I am was only testing one camera at the moment. Need to open up the second camera later to test multi streaming. I kind of figured it wasn't resource intensive but did the test since it wasn't much effort to change the settings and restart the VM.
